### PR TITLE
Add nearcaching hook to map-transaction-log-record 

### LINKED ISSUE #16577: Near Cache Invalidation not correct when used with TransactionalMap
**Describe the bug**
When using a transactional map from a cache with a near cache, the cache invalidation event occurs when the transactionalMap.put is called. As a result, the entry gets invalidated BEFORE the change is committed to the map.  Any readers calling get on the entry cause the old entry to be stored in the near cache, even though the correct entry gets stored in the underlying map.

**Expected behaviour**
I would expect that during the call to TransactionContext.commitTransaction();, after the entries have been committed to the underlying map, only then does the cache invalidation message get sent.

**To Reproduce**
Run the following code, and notice the last System.out does not output the expected value.  This is a simple test case after we found the issue in our code as soon as we enabled near cache for one of the caches.

```
public static void main(String[] args) {
        Config cfg = new Config("one");
        MapConfig cacheConfig = cfg.getMapConfig("cache");
        NearCacheConfig nearCacheConfig = new NearCacheConfig(1000, 1000, true, InMemoryFormat.OBJECT);
        cacheConfig.setNearCacheConfig(nearCacheConfig);
        nearCacheConfig.setCacheLocalEntries(true);

        // create Hazelcast instance
        final HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(cfg);
        IMap map = hcInstance.getMap("cache");
        map.put("key", "value"); // Initialise with a value
        System.out.println("Value prior (should be 'value') was " + map.get("key"));

        TransactionOptions opts = new TransactionOptions();
        opts.setTransactionType(TransactionOptions.TransactionType.TWO_PHASE);
        TransactionContext ctx = hcInstance.newTransactionContext(opts);
        ctx.beginTransaction(); // being the transaction
        TransactionalMap transactionalMap = ctx.getMap("cache");
        System.out.println("Value prior in transaction map (should be 'value') was " + transactionalMap.get("key"));
        transactionalMap.put("key", "newValue");
        System.out.println("Value post put in transaction map (should be 'newValue') was " + transactionalMap.get("key"));
        System.out.println("Value post put in map (should be 'value') was " + map.get("key"));
        ctx.commitTransaction();
        System.out.println("Value post commit in map (should be 'newValue') was " + map.get("key"));
        System.exit(0);
    }
```

**Additional context**
This has been tested on 3.11.6 and 3.12.5 - both exhibit the same behaviour.

### PR DESCRIPTION
**Pull request by ahmetmircik**
closes #16577
Backport of: https://github.com/hazelcast/hazelcast/pull/16579

### CODE REVIEW

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java [hunk 1]
```diff
@@ -175,18 +175,41 @@ public void commit() throws TransactionException, IllegalStateException {
         }
     }
 
-    public void commitAsync(ExecutionCallback callback) {
+    public void commitAsync(final ExecutionCallback callback) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java]: Shouldn't we add these too?
https://github.com/hazelcast/hazelcast/pull/16579/files#diff-0c104d8009708b16b9092edfd2438fa1R173-R176

File: hazelcast/src/main/java/com/hazelcast/internal/nearcache/NearCachingHook.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        public void onRemoteCallSuccess() {
+        }
+
+        @Override
+        public void onRemoteCallFailure() {
+
+        }
+    };
+
+    void beforeRemoteCall(K key, Data keyData, V value, Data valueData);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/internal/nearcache/NearCachingHook.java]: How come you don't use `@Nullable` here?

File: hazelcast/src/main/java/com/hazelcast/map/impl/tx/MapTransactionLogRecord.java [hunk 3]
```diff
@@ -39,10 +40,15 @@
     private String ownerUuid;
     private Operation op;
 
+    private transient NearCachingHook nearCachingHook = NearCachingHook.EMPTY_HOOK;
+
     public MapTransactionLogRecord() {
     }
 
-    public MapTransactionLogRecord(String name, Data key, int partitionId, Operation op, long version, String ownerUuid) {
+    public MapTransactionLogRecord(String name, Data key, int partitionId,
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/tx/MapTransactionLogRecord.java]: Can the parameter be`@Nullable`?

File: hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java]: I found the same is missing in https://github.com/hazelcast/hazelcast/pull/16618 and https://github.com/hazelcast/hazelcast/pull/16617.

File: hazelcast/src/main/java/com/hazelcast/internal/nearcache/NearCachingHook.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/nearcache/NearCachingHook.java]: It wasn't common in 3.12 (java 6) using those, no real reason.

File: hazelcast/src/main/java/com/hazelcast/map/impl/tx/MapTransactionLogRecord.java (hunk 3 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/tx/MapTransactionLogRecord.java]: same

File: hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java]: didn't get this?

**Comment by ahmetmircik** [comment]
run-lab-run

**Pull request merged**